src/mds/MDBalancer.cc

```diff
@@ -169,6 +171,28 @@ mds_load_t MDBalancer::get_load(utime_t now)
   return load;
 }
 
+int MDBalancer::localize_balancer(string const balancer)
+{
+  int64_t pool_id = mds->mdsmap->get_metadata_pool();
+  string fname = "/tmp/" + balancer;
+
+  dout(15) << "looking for balancer=" << balancer << " in RADOS pool_id=" << pool_id << dendl;
+  object_t oid = object_t(balancer);
+  object_locator_t oloc(pool_id);
+  bufferlist data;
+  C_SaferCond waiter;
+  mds->objecter->read(oid, oloc, 0, 0, CEPH_NOSNAP, &data, 0, &waiter);
```

Comment:
Ouch, this is a blocking OSD read that's done from inside the mds_lock when this gets called from mantle_prep_rebalance.

I think it would be better to kick off the read when handling the mdsmap (if the balancer has changed), and populate the script in the background when the read completes.  If mantle_prep_rebalance runs before that read has completed it can either no-op or fall back to the native balancer.